### PR TITLE
refactor output perturbation mechanisms

### LINKED ISSUE #1307: reduce use of traits in noise perturbation mechanisms
The implementation of noise perturbation mechanisms can be simplified to make them easier to prove.

### LINKED ISSUE #1308: remove `_base_` variations of output perturbation mechanisms
We should always use the fused constructors, at this point. For simplicity, the variations should be hidden from the user as well.

### LINKED ISSUE #1309: relaxation factors on rounding in float perturbation mechanisms should scale with the dimensionality

### PR DESCRIPTION
**Pull request by Shoeboxam**
Closes #1307
Closes #1308
Closes #1309

* separate constructors for each combination of scalar/vector x integer/float x gaussian/laplace. 
* removed all constructors from the public API except for `make_laplace`, `make_gaussian` and `make_geometric`
* `make_laplace` and `make_gaussian` now have an argument `k` for float noise granularity
* `make_geometric` has `bounds` instead of `k` for constant-time execution
* updated all docs/tests to remove references to the variations of `base_` constructors
* increased relaxation term for vector-valued float mechanisms

### CODE REVIEW

**Comment by Shoeboxam** [comment]
* **#1321** <a href="https://app.graphite.com/github/pr/opendp/opendp/1321?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1320** <a href="https://app.graphite.com/github/pr/opendp/opendp/1320?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1319** <a href="https://app.graphite.com/github/pr/opendp/opendp/1319?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1318** <a href="https://app.graphite.com/github/pr/opendp/opendp/1318?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/opendp/opendp/1318?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1317** <a href="https://app.graphite.com/github/pr/opendp/opendp/1317?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1326** <a href="https://app.graphite.com/github/pr/opendp/opendp/1326?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#1315](https://github.com/opendp/opendp/pull/1315) <a href="https://app.graphite.com/github/pr/opendp/opendp/1315?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#1327](https://github.com/opendp/opendp/pull/1327) <a href="https://app.graphite.com/github/pr/opendp/opendp/1327?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1325** <a href="https://app.graphite.com/github/pr/opendp/opendp/1325?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by mccalluc — COMMENTED** [review]
I think the biggest question for me is the scope: After this, should there be any `base_` functions remaining? A few places in here seem like they were overlooked, but not sure.

File: R/opendp/NAMESPACE
Comment (Shoeboxam) [reply R/opendp/NAMESPACE:56]: I was going to handle it in a separate PR. There's some ambiguity between PSRNs and this discrete approach, I've recently had a couple changed circumstances that has made me cool on the PSRNs.

File: R/opendp/tests/testthat/test-measurements.R
Comment (Shoeboxam) [reply R/opendp/tests/testthat/test-measurements.R:143]: It was for parity with python tests, but we don't really have parity anyways.

File: R/opendp/tests/testthat/test-measurements.R
Comment (Shoeboxam) [reply R/opendp/tests/testthat/test-measurements.R:181]: I've got a VSCode task that runs all tests in one file:

```json
         {
            "label": "R/ test file",
            "type": "shell",
            "group": "build",
            "options": {
                "cwd": "./R/opendp/",
            },
            "command": "Rscript -e 'devtools::load_all(recompile=TRUE); library(testthat); test_file(\"${file}\", reporter = c(\"location\", \"progress\"))'",
            "presentation": {
                "clear": true
            }
        },
```

Also, testing splits output by file. So if we want to group differently, we should adjust the files themselves.

File: docs/source/quickstart.rst
Comment (Shoeboxam) [reply docs/source/quickstart.rst:60]: I've started using t_ and m_ prefixes in variable names to denote transformation or measurement. Will change here.

File: docs/source/user/measurements.rst
Comment (Shoeboxam) [reply docs/source/user/measurements.rst:58]: I reduced it to two rows each, generic over the dtype, because atom pairs only with absolute and vector only with l1.

File: rust/src/combinators/laplace_delay/mod.rs
Comment (Shoeboxam) [reply rust/src/combinators/laplace_delay/mod.rs:1]: This file should not be in here, I must have made an edit to a shadow file and it got added. Thanks for noticing.

File: rust/src/measurements/alp/mod.rs
Comment (Shoeboxam) [reply rust/src/measurements/alp/mod.rs:215]: Hmm, I don't see it. Maybe I already fixed it since you wrote the review?

File: rust/src/measurements/laplace/mod.rs
Comment (Shoeboxam) [reply rust/src/measurements/laplace/mod.rs:104]: The timing is interesting, now the linear time sampler is always slower. I might as well simplify/remove it, and only make it accessible from the geometric constructor when bounds are set.

Comment (Shoeboxam) [reply rust/src/measurements/laplace/mod.rs:104]: Simplified.

File: docs/source/quickstart.R
Comment (mccalluc) [inline docs/source/quickstart.R:7]: Particularly on the examples, I'd really prefer to to spell out the variable names. This is their first introduction to OpenDP.

File: docs/source/quickstart.rst
Comment (mccalluc) [inline docs/source/quickstart.rst:61]: Again, prefer to avoid abbreviations here.

File: docs/source/quickstart.rst
Comment (mccalluc) [inline docs/source/quickstart.rst:74]:
```suggestion
* Invoking the ``m_laplace`` measurement on a value to get a DP release.
```
or 
```suggestion
* Invoking ``laplace_measurement`` on a value to get a DP release.
```

This should match the name, whatever the name is.

File: rust/src/measurements/alp/mod.rs
Comment (mccalluc) [reply rust/src/measurements/alp/mod.rs:215]: and I don't see it either now. I must have been out of date.

File: rust/src/measurements/gaussian/float/mod.rs
Comment (mccalluc) [inline rust/src/measurements/gaussian/float/mod.rs:56]:
```suggestion
/// Make a Measurement that adds noise from the Gaussian(`scale`) distribution to the vector-valued input.
```
Make the description analogous to that for ` make_scalar_float_gaussian`.

**Review by mccalluc — DISMISSED** [review]
A couple of thoughts on documentation, but nothing that should hold this up. In quickstart, I'll probably make it a longer name after this is merged to main, and main is merged to the docs branch, unless you really like the `m_` convention.

Approving, and then I'll see if I get set it to auto-merge.

**Review by mccalluc — COMMENTED** [review]
Some comments; Mostly things that seemed odd, but I could just be confused.

Github has the option to look at changes since last review, but it doesn't work here, maybe because of rebases or something like that. I don't know whether there is a workflow that would enable seeing just the most recent changes, without making everything else more difficult.

One higher level question: I think by default we no longer show the Rust stack trace in python, but Rust error messages will come through? So we should make sure the error messages will make sense to users of the language bindings? I wasn't seeing any error messages that would be too cryptic, but I wanted to check my understanding.

File: rust/src/measurements/gaussian/mod.rs
Comment (mccalluc) [reply rust/src/measurements/gaussian/mod.rs:144]: Or maybe this is fine, as long as we're not showing a scary rust stacktrace?

File: rust/src/measurements/gaussian/mod.rs
Comment (mccalluc) [reply rust/src/measurements/gaussian/mod.rs:170]: ignore

File: rust/src/measurements/geometric/integer/mod.rs
Comment (Shoeboxam) [reply rust/src/measurements/geometric/integer/mod.rs:96]: Bounds are optional-- if they are not specified, then the check does not need to be run.

File: rust/src/measurements/gaussian/mod.rs
Comment (Shoeboxam) [reply rust/src/measurements/gaussian/mod.rs:144]: There are many settings where errors from rust trickle up- when impossible pairings of bounds are set, domains or metrics mismatch, or the scale is negative. Maybe I'm misunderstanding the complaint.

File: docs/source/getting-started/examples/histograms.ipynb
Comment (Shoeboxam) [reply docs/source/getting-started/examples/histograms.ipynb:91]: I think it's right as-is-- the category set is a set of atomic values, not sets themselves.

File: R/opendp/tests/testthat/test-measurements.R
Comment (Shoeboxam) [reply R/opendp/tests/testthat/test-measurements.R:70]: Ideally, the substantive tests are all handled in Rust, and the tests from bindings languages are stripped down to be as thin as possible, so that they just test the bindings. This should make it possible to automate test generation as well. Codegen for tests is a low priority, but this policy toward testing will make it less of a lift to automate in the future. 

Going by this approach, then yes, new tests for functionality in Rust should match each other in R and Python, be as thin as possible in R and Python, and be concentrated more in Rust.

**Comment by Shoeboxam** [comment]
The three commits starting from "address review comments" are the new things- you can select that subset of commits. Given that there was a rebase, I could have mixed changes into prior commits anywhere along the rebase, which is why the button didn't work. In this case, I think I had to rebase, as there were conflicts, but I usually try not to rebase between reviews so as to avoid this issue.

**Review by mccalluc — APPROVED** [review]
Thanks for explanations -- Looks good!

**Pull request merged**
